cuda/base/kernel_launch.cuh

```diff
@@ -0,0 +1,243 @@
+
+template <typename ValueType>
+struct to_device_type_impl<matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<cuda_type<ValueType>>;
+    static type map_to_device(matrix::Dense<ValueType> *mtx)
+    {
+        return {as_cuda_type(mtx->get_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<const cuda_type<ValueType>>;
+    static type map_to_device(const matrix::Dense<ValueType> *mtx)
+    {
+        return {as_cuda_type(mtx->get_const_values()), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<Array<ValueType> &> {
+    using type = cuda_type<ValueType> *;
+    static type map_to_device(Array<ValueType> &array)
+    {
+        return as_cuda_type(array.get_data());
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const Array<ValueType> &> {
+    using type = const cuda_type<ValueType> *;
+    static type map_to_device(const Array<ValueType> &array)
+    {
+        return as_cuda_type(array.get_const_data());
+    }
+};
+
+
+template <typename T>
+typename to_device_type_impl<T>::type map_to_device(T &&param)
+{
+    return to_device_type_impl<T>::map_to_device(param);
+}
+
+
+template <typename ValueType>
+struct compact_dense_wrapper {
+    ValueType *data;
+};
+
+
+template <typename T>
+struct device_unpack_2d_impl {
+    using type = T;
+    static __device__ __forceinline__ type unpack(T param, size_type)
+    {
+        return param;
+    }
+};
+
+template <typename ValueType>
+struct device_unpack_2d_impl<compact_dense_wrapper<ValueType>> {
+    using type = matrix_accessor<ValueType>;
+    static __device__ __forceinline__ type
+    unpack(compact_dense_wrapper<ValueType> param, size_type num_cols)
+    {
+        return {param.data, num_cols};
+    }
+};
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_1d(
+    size_type size, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    if (tidx >= size) {
+        return;
+    }
+    fn(tidx, args...);
+}
+
+
+template <typename KernelFunction, typename... KernelArgs>
+__global__ __launch_bounds__(default_block_size) void generic_kernel_2d(
+    size_type rows, size_type cols, KernelFunction fn, KernelArgs... args)
+{
+    auto tidx = thread::get_thread_id_flat();
+    auto col = tidx % cols;
+    auto row = tidx / cols;
+    if (row >= rows) {
+        return;
+    }
+    fn(row, col, device_unpack_2d_impl<KernelArgs>::unpack(args, cols)...);
+}
+
+
+}  // namespace cuda
+}  // namespace kernels
+
+
+template <typename ValueType>
+kernels::cuda::compact_dense_wrapper<kernels::cuda::cuda_type<ValueType>>
+compact(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {kernels::cuda::as_cuda_type(mtx->get_values())};
+}
```

Comment:
This function needs a more descriptive name, or a separate namespace (maybe even both since this is in the top-level namespace). I still don't quite understand why it is necessary to have such a function. Is the intention to avoid an additional parameter for a stride that is known to be `size[1]`?

Reply:
Yes, it avoids having a separate stride parameter for each parameter, which might become significant for solver kernels with many parameters.

Reply:
In that case, I would remove these functions or, at the very least, move them into the `dense` namespace (we should not pollute the `gko` namespace with something only used for `Dense` matrices).
Personally, I would remove them since they are purely for niche optimization purposes and they make the code a lot harder to read.

Reply:
What do you think about the new `solver::rowvector` and `solver::default_stride` wrappers instead?

Reply:
I like them, and since they are in a different namespace, I am fine with them.
One nit: Why `rowvector` and not `row_vector`? I feel like it is easier to read.

Reply:
Yes, that's probably better